File: packages/js/onnx2gguf/src/tokenizer.ts
/* eslint-disable */
export function extractTokenizerMetadata(
  tokenizerJsonStr: string | null = null,
  vocabSize: number = 0,
): Record<string, ReturnType<typeof JSON.parse>> {
  const meta: Record<string, ReturnType<typeof JSON.parse>> = {};

  if (!tokenizerJsonStr) {
    meta['tokenizer.ggml.model'] = 'llama';
    const tokens = [];
    const size = vocabSize > 0 ? vocabSize : 2;
    for (let i = 0; i < size; i++) {
      tokens.push(vocabSize > 0 ? `[TOKEN_${i}]` : i === 0 ? '<s>' : '</s>');
    }
    meta['tokenizer.ggml.tokens'] = tokens;
    meta['tokenizer.ggml.scores'] = Array(size).fill(0.0);
    meta['tokenizer.ggml.token_type'] = Array(size).fill(1);
    meta['tokenizer.ggml.bos_token_id'] = 0;
    meta['tokenizer.ggml.eos_token_id'] = 1;
    meta['tokenizer.ggml.unknown_token_id'] = 0;
    meta['tokenizer.ggml.padding_token_id'] = 0;
    meta['tokenizer.ggml.separator_token_id'] = 0;
    meta['tokenizer.ggml.add_bos_token'] = true;
    meta['tokenizer.ggml.add_eos_token'] = false;
    meta['tokenizer.chat_template'] = '';
    return meta;
  }

  let t: ReturnType<typeof JSON.parse>;
  try {
    t = JSON.parse(tokenizerJsonStr);
  } catch (e) {
    meta['tokenizer.ggml.model'] = 'llama';
    return meta;
  }

  const model = t.model || {};
  const modelType = model.type || 'BPE';

  if (modelType === 'BPE') {
    meta['tokenizer.ggml.model'] = 'gpt2';
  } else if (modelType === 'Unigram') {
    meta['tokenizer.ggml.model'] = 'llama';
  } else {
    meta['tokenizer.ggml.model'] = 'llama';
  }

  const vocab = model.vocab || {};
  if (typeof vocab === 'object' && vocab !== null && !Array.isArray(vocab)) {
    const sortedVocab = Object.entries(vocab).sort(
      (a: ReturnType<typeof JSON.parse>, b: ReturnType<typeof JSON.parse>) => a[1] - b[1],
    );
    let tokens = sortedVocab.map((x) => x[0]);

    if (vocabSize > 0 && tokens.length !== vocabSize) {
      if (tokens.length < vocabSize) {
        const diff = vocabSize - tokens.length;
        const offset = tokens.length;
        for (let i = 0; i < diff; i++) {
          tokens.push(`[DUMMY_${offset + i}]`);
        }
      } else {
        tokens = tokens.slice(0, vocabSize);
      }
    }

    meta['tokenizer.ggml.tokens'] = tokens;
    meta['tokenizer.ggml.scores'] = Array(tokens.length).fill(0.0);
    meta['tokenizer.ggml.token_type'] = Array(tokens.length).fill(1);
  }

  const merges = model.merges || [];
  if (merges.length > 0) {
    meta['tokenizer.ggml.merges'] = merges;
  }

  meta['tokenizer.ggml.bos_token_id'] =
    t.added_tokens && t.added_tokens.length > 0 ? t.added_tokens[0].id || 0 : 0;
  meta['tokenizer.ggml.eos_token_id'] =
    t.added_tokens && t.added_tokens.length > 0
      ? t.added_tokens[t.added_tokens.length - 1].id || 1
      : 1;
  meta['tokenizer.ggml.unknown_token_id'] = 0;
  meta['tokenizer.ggml.padding_token_id'] = 0;
  meta['tokenizer.ggml.separator_token_id'] = 0;
  meta['tokenizer.ggml.add_bos_token'] = true;
  meta['tokenizer.ggml.add_eos_token'] = false;
  meta['tokenizer.chat_template'] = t.chat_template || '';

  return meta;
}

File: packages/python/onnx9000-onnx2gguf/src/onnx9000/onnx2gguf/tokenizer.py
import json
import base64
from typing import Any


def extract_tokenizer_metadata(
    tokenizer_json_str: str = None, vocab_size: int = 0
) -> dict[str, Any]:
    meta = {}

    if not tokenizer_json_str:
        # 89. Generate a dummy tokenizer if no tokenizer metadata is provided
        meta["tokenizer.ggml.model"] = "llama"
        tokens = [f"[TOKEN_{i}]" for i in range(vocab_size)] if vocab_size > 0 else ["<s>", "</s>"]
        if vocab_size == 0:
            vocab_size = 2
        meta["tokenizer.ggml.tokens"] = tokens
        meta["tokenizer.ggml.scores"] = [0.0] * vocab_size
        meta["tokenizer.ggml.token_type"] = [1] * vocab_size
        meta["tokenizer.ggml.bos_token_id"] = 0
        meta["tokenizer.ggml.eos_token_id"] = 1
        meta["tokenizer.ggml.unknown_token_id"] = 0
        meta["tokenizer.ggml.padding_token_id"] = 0
        meta["tokenizer.ggml.separator_token_id"] = 0
        meta["tokenizer.ggml.add_bos_token"] = True
        meta["tokenizer.ggml.add_eos_token"] = False
        meta["tokenizer.chat_template"] = ""
        return meta

    try:
        t = json.loads(tokenizer_json_str)
    except json.JSONDecodeError:
        # Maybe it's a base64 sentencepiece model
        meta["tokenizer.ggml.model"] = "llama"
        return meta

    model = t.get("model", {})
    model_type = model.get("type", "BPE")

    if model_type == "BPE":
        meta["tokenizer.ggml.model"] = "gpt2"
    elif model_type == "Unigram":
        meta["tokenizer.ggml.model"] = "llama"
    else:
        meta["tokenizer.ggml.model"] = "llama"

    vocab = model.get("vocab", {})
    if isinstance(vocab, dict):
        sorted_vocab = sorted(vocab.items(), key=lambda x: x[1])
        tokens = [k for k, v in sorted_vocab]

        # Validate length
        if vocab_size > 0 and len(tokens) != vocab_size:
            # 90. Validate length matches vocab_size exactly
            # If not, pad it or truncate it
            if len(tokens) < vocab_size:
                tokens.extend([f"[DUMMY_{i}]" for i in range(len(tokens), vocab_size)])
            else:
                tokens = tokens[:vocab_size]

        meta["tokenizer.ggml.tokens"] = tokens
        meta["tokenizer.ggml.scores"] = [0.0] * len(tokens)
        meta["tokenizer.ggml.token_type"] = [1] * len(tokens)

    merges = model.get("merges", [])
    if merges:
        meta["tokenizer.ggml.merges"] = merges

    meta["tokenizer.ggml.bos_token_id"] = (
        t.get("added_tokens", [{}])[0].get("id", 0) if t.get("added_tokens") else 0
    )
    meta["tokenizer.ggml.eos_token_id"] = (
        t.get("added_tokens", [{}])[-1].get("id", 1) if t.get("added_tokens") else 1
    )
    meta["tokenizer.ggml.unknown_token_id"] = 0
    meta["tokenizer.ggml.padding_token_id"] = 0
    meta["tokenizer.ggml.separator_token_id"] = 0
    meta["tokenizer.ggml.add_bos_token"] = True
    meta["tokenizer.ggml.add_eos_token"] = False

    # Optional chat template
    meta["tokenizer.chat_template"] = t.get("chat_template", "")

    return meta
